_source/_posts/2023-01-24-jakartaee-auth0.md

```diff
@@ -0,0 +1,545 @@
+    │   └── openid.properties  // OpenID config properties
+    └── webapp
+        └── WEB-INF
+            └── jboss-web.xml  // Configures context root to '/'
+```
+
+When the application loads, the OpenID properties are loaded from `openid.properties` by the `OpenIDConfig` class. These values are used by the `JwtFilter` to create the class that verifies JSON Web Tokens. These properties are also used by the `ProtectedServlet` in the `@OpenIdAuthenticationMechanismDefinition` annotation to configure OIDC.
+
+The `jboss-web.xml` file is simply used to change the context root to `/`.
+
+Neither of the files in the `META-INF` directory seem to be required for the application to function. The `beans.xml` file explicitly enables CDI / dependency injection. However, this would also be done implicitly by the use of the annotations. The `MANIFEST.MF` file defines some provided runtime dependencies related to CDI. Perhaps in some runtime environments it would be necessary to include this file, but it seems unneeded when running locally with WilfFly.
+
+## Create an Auth0 API and OIDC application
+
+If you have not already, install the [Auth0 CLI](https://github.com/auth0/auth0-cli) and run `auth0 login` in a terminal.
+
+```bash
+Waiting for the login to complete in the browser... done
+
+ ▸    Successfully logged in.
+ ▸    Tenant: dev-0xb84jzp.us.auth0.com
+```
+
+Take note of the domain listed as the tenet. This is your Auth0 domain. If you need to find it again later, you can use `auth0 tenants list`.
+
+You need to create an API on Auth0. The Auth0 API is what exposes identity functionality for all authentication and authorization protocols, such as OpenID Connect and OAuth. **Without the API, Auth0 servers will return an opaque token that the Java application will not be able to verify.** This can lead to some cryptic error messages and some lost time. 
+
+Use the following command to create a custom Auth0 API named `myapi` with the identifier `http://my-api`.
+
+```bash
+auth0 apis create -n myapi --identifier http://my-api
+```
+
+Just press enter three times to accept the default values for scopes, token lifetime, and to allow offline access. The scopes here refer to custom scopes, not the standard scopes (email, profile, and openid) that you will need for OIDC and OAuth.
+
+Now use the Auth0 CLI to create an OpenID Connect (OIDC) application.  From the project base directory, run the following.
+
+```bash
+auth0 apps create
+```
+
+Use the following values:
+
+- **Name**: `javartaee-demo`
+
+- **Description**: whatever you like, or leave blank
+- **Type**: `Regular Web Application`
+- **Callback URLs**: `http://localhost:8080/callback`
+- **Allowed Logout URLs**: `http://localhost:8080`
+
+The console output shows you the Auth0 domain and the OIDC client ID. However, you also need the client secret, which you have to get by logging into Auth0. Type the following:
+
+```bash
+auth0 apps open
```

Comment:
With the 1.0 version of the Auth0 CLI, you can use `--reveal-secrets` to see the client secret!

https://github.com/auth0/auth0-cli/issues/488#issuecomment-1404051864